server/server.go

```diff
@@ -65,59 +175,180 @@ func readFormData(r *http.Request) string {
+		case tar.TypeDir:
+			fmt.Println("Dir ", filename)
+			err := os.MkdirAll(filepath.Join(dest, filename), os.FileMode(header.Mode)) // or use 0755 if you prefer
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+		case tar.TypeReg:
+			fmt.Println("File ", filename)
+			err := os.MkdirAll(filepath.Join(dest, filepath.Dir(filename)), os.FileMode(header.Mode))
+			writer, err := os.Create(filepath.Join(dest, filename))
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			_, err = io.Copy(writer, unTarred)
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			err = os.Chmod(filepath.Join(dest, filename), os.FileMode(header.Mode))
+
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			writer.Close()
+		default:
+			log.Println("unable to unTar type : ", header.Typeflag)
+			return errResponse
 		}
 	}
+	return ""
+}
+
+// storeUnzippedFiles stores unzipped files to 'assignments' directory.
+func storeUnzippedFiles(unZipped *zip.Reader) string {
+	dest := filepath.Join(constants.AssignmentsDir, assignTestingInfo.RootDir)
 
+	errorResponse := `"Unzip Error":"Error in unzipping uploaded file"`
+
+	for _, file := range unZipped.File {
+		fPath := filepath.Join(dest, file.Name)
+
+		if !strings.HasPrefix(fPath, filepath.Clean(dest)+string(os.PathSeparator)) {
+			log.Println("unzip error: illegal filepath")
+			return errorResponse
+		}
+
+		if file.FileInfo().IsDir() {
+			err := os.MkdirAll(fPath, os.ModePerm)
+			if err != nil {
+				log.Println("unzip error: ", err)
+				return errorResponse
+			}
+			continue
+		}
+
+		err := os.MkdirAll(filepath.Dir(fPath), os.ModePerm)
+		if err != nil {
+			log.Println("unzip error: ", err)
+			return errorResponse
+		}
+
+		outFile, err := os.OpenFile(fPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
+		if err != nil {
+			log.Println("unzip error: ", err)
+			return errorResponse
+		}
+
+		fileReader, err := file.Open()
+		if err != nil {
+			log.Println("unzip error: ", err)
+			return errorResponse
+		}
+
+		_, err = io.Copy(outFile, fileReader)
+
+		// Close the file without defer to close before next iteration of loop.
+		outFile.Close()
+		fileReader.Close()
+
+		if err != nil {
+			log.Println("unzip error: ", err)
+			return errorResponse
+		}
+
+	}
 	return ""
 }
 
+// listenAndServe listens to requests on the given port number.
 func listenAndServe(wg *sync.WaitGroup, port string) {
 	defer wg.Done()
 
 	log.Printf("** Service Started on Port " + port + " **")
 	if err := http.ListenAndServe(":"+port, nil); err != nil {
-		log.Fatal(err)
+		log.Println(err)
 	}
 }
+
+// StartServer starts service at given port.
 func StartServer(port string) {
 
 	var wg sync.WaitGroup
 
-	// Start service at given port.
 	http.HandleFunc("/upload", upload)
-
+	http.HandleFunc("/buildRun", buildRun)
 	wg.Add(1)
 	go listenAndServe(&wg, port)
```

Comment:
So, when does the server shutdown?
How to shutdown the server?

Reply:
With Ctrl + C. But I will raise an issue to implement graceful shutdown